### PR TITLE
community standards: add contributing guideline

### PR DESCRIPTION
**Pull request by 4rgon4ut**
This PR adds CONTRIBUTING.md to the root.
Related discussion issue #337.

### CODE REVIEW

**Comment by 4rgon4ut** [comment]
Clippy warnings fixed in https://github.com/maticnetwork/miden/pull/364/commits/6b52bf44b3bca994d333dcc252db51eacc3a4983 PR, waiting for merge to rebase.

File: CONTRIBUTING.md
Comment (bobbinth) [inline CONTRIBUTING.md:17]: For w/e reason this link takes me to `https://docs.github.com/en`. Should we use a different link? e.g.: https://docs.github.com/get-started/quickstart/github-flow

File: CONTRIBUTING.md
Comment (bobbinth) [inline CONTRIBUTING.md:65]: nit: missing space after "scope". Should be "... and scope (if possible) ...".

File: CONTRIBUTING.md
Comment (bobbinth) [inline CONTRIBUTING.md:74]: Should we include the actual command which people can run? For example, something like:
```
cargo fix --allow-staged --allow-dirty --all-targets && cargo fmt && cargo clippy
```
Or maybe even:
```
cargo fix --allow-staged --allow-dirty --all-targets --all-features; cargo fmt; cargo clippy --workspace --all-targets --all-features -- -D warnings
```

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks great! Thank you again for putting this together. I left a couple of small comments inline.

File: CONTRIBUTING.md
Comment (grjte) [reply CONTRIBUTING.md:74]: Agreed! I was going to suggest this as well.

File: CONTRIBUTING.md
Comment (grjte) [reply CONTRIBUTING.md:65]: nit: there are a couple typos here, so I'd suggest a minor edit to the following - 

> Commit messages should be written in a short, descriptive manner and be prefixed with tags for the change type and scope (if possible) according to the [semantic commit](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) scheme."

I think it might also be worth adding an example, such as:

> For example, a new change to the AIR crate might have the following message: "feat(air): add constraints for the decoder"

File: CONTRIBUTING.md
Comment (grjte) [inline CONTRIBUTING.md:10]: nit: "Taking part in ..."

File: CONTRIBUTING.md
Comment (grjte) [inline CONTRIBUTING.md:11]: nit: It feels like this should have a link as well, since all the others do. Maybe we should link to the pull requests page? Not a big deal though, so whatever you prefer is fine.

File: CONTRIBUTING.md
Comment (grjte) [reply CONTRIBUTING.md:17]: I agree with this. I also wonder if we should include a link to the quickstart on forking (https://docs.github.com/en/get-started/quickstart/fork-a-repo). It includes info about setting the upstream & staying in sync, which provides more detail related to the rebasing discussed later. We could amend the text to say "... so all code changes happen through pull requests from a [forked repo](https://docs.github.com/en/get-started/quickstart/fork-a-repo)."

I know this is fairly obvious, but forking isn't mentioned in the other github flow link, and you never know when we'll be someone's first open source project :)

Whatever you guys think, though

File: CONTRIBUTING.md
Comment (grjte) [inline CONTRIBUTING.md:23]: nit: This is missing a word - I'd say "... Every branch with a fix/feature must be..."

File: CONTRIBUTING.md
Comment (grjte) [inline CONTRIBUTING.md:25]: nit: There's a typo in "description". I would also say "The branch name..." instead of just "Branch name..."

File: CONTRIBUTING.md
Comment (grjte) [inline CONTRIBUTING.md:27]: nit: I would probably type out "something" fully here instead of "smth". I would also say "...the branch name..." again instead of just "...branch name..."

File: CONTRIBUTING.md
Comment (grjte) [inline CONTRIBUTING.md:82]: nit: maybe this line should remind them to follow the documentation convention as well, like "Documentation/comments updated for all changes according to our documentation convention."

File: CONTRIBUTING.md
Comment (grjte) [inline CONTRIBUTING.md:69]: very minor nit: since the sample below isn't 100 characters, maybe this should be slightly rephrased for clarity to:

> For code sections, we use code separators like the following to a width of 100 characters:

**Review by grjte — CHANGES_REQUESTED** [review]
Thank you @4rgon4ut! I left several trivial nits inline, including some optional things that you can leave off if you disagree, but I think this is really well done, especially all the visuals & examples

File: CONTRIBUTING.md
Comment (grjte) [reply CONTRIBUTING.md:17]: I agree with this. I also wonder if we should include a link to the quickstart on forking (https://docs.github.com/en/get-started/quickstart/fork-a-repo). It includes info about setting the upstream & staying in sync, which provides more detail related to the rebasing discussed later. We could amend the text to say "... so all code changes happen through pull requests from a [forked repo](https://docs.github.com/en/get-started/quickstart/fork-a-repo)."

I know this is fairly obvious, but forking isn't mentioned in the other github flow link, and you never know when we'll be someone's first open source project :)

Whatever you guys think, though

File: CONTRIBUTING.md
Comment (grjte) [reply CONTRIBUTING.md:65]: nit: there are a couple typos here, so I'd suggest a minor edit to the following - 

> Commit messages should be written in a short, descriptive manner and be prefixed with tags for the change type and scope (if possible) according to the [semantic commit](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) scheme."

I think it might also be worth adding an example, such as:

> For example, a new change to the AIR crate might have the following message: "feat(air): add constraints for the decoder"

File: CONTRIBUTING.md
Comment (4rgon4ut) [reply CONTRIBUTING.md]: Github flow link updated and forking guide link added. I think that links with additional information always a good idea in guides. Especially when we are not losing anything as a tradeoff source.

Comment (4rgon4ut) [reply CONTRIBUTING.md]: Github flow link updated and forking guide link added. I think that links with additional information always a good idea in guides. Especially when we are not losing anything as a tradeoff source.

File: CONTRIBUTING.md
Comment (4rgon4ut) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md
Comment (4rgon4ut) [reply CONTRIBUTING.md]: Done in a way that @grjte proposed.

File: CONTRIBUTING.md
Comment (4rgon4ut) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md
Comment (4rgon4ut) [reply CONTRIBUTING.md]: I'm thought about that too. We have no link to really fit this point, maybe issue creation page link is closest one, because any fix will be firstly proposed through new issue. But having 3/4 bullet points with link to issue creation page will look strange. So, I'm added link to the pull requests page for visual consistency.

File: CONTRIBUTING.md
Comment (4rgon4ut) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md
Comment (4rgon4ut) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md
Comment (4rgon4ut) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md
Comment (4rgon4ut) [reply CONTRIBUTING.md]: Done.

**Comment by 4rgon4ut** [comment]
@grjte @bobbinth Thank you both for suggestions and review! I added all requested changes, I think in context of community standards the last word always on core team members :)

**Review by bobbinth — APPROVED** [review]
All looks good to me! Thank you for working on this!

**Pull request merged**
